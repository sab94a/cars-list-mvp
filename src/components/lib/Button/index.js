// @flow

import React from 'react';
import type { Button as Props} from 'components/lib/types'
import styles from './index.module.scss';

const Button = ({ children, ...rest }: Props) => (
    <button {...rest } className={ styles.root } >{ children }</button>
);

export default React.memo<Props>(Button)
