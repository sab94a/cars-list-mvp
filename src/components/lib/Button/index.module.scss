@import 'components/lib/variables.scss';

.root {
    border: 0;
    background: $primary;
    color: #fff;
    width: $button-width;
    font-size: $font-size-sm;
    padding: $offset-sm;
    border-radius: 2px;
    cursor: pointer;

    &:hover, &:focus {
        background: $primary-pressed;
    }
}