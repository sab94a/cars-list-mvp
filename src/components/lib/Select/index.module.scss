@import 'components/lib/variables.scss';

.root {
    position: relative;
    font-size: $font-size-sm;
    outline: none;
    width: $button-width * 2;
    text-align: left;
}

.list {
    position: absolute;
    top: 100%;
    left: 0;
    padding: 0;
    margin: $offset-sm 0 0 0;
    min-width: 100%;
    border-radius: $border-width-sm;
    border: $border-width-sm solid $light;
    z-index: 1;
}

.selected {
    position: relative;
    border: $border-width-sm solid $light;

    &:after {
        content: '';
        display: block;
        position: absolute;
        right: $offset-sm;
        top: 50%;
        margin-top: -$offset-sm / 2;
        width: 0;
        border: 7px solid transparent;	
        border-top: 10px solid $light;
    }

    &.open:after {
        border: 7px solid transparent;	
        margin-top: -13px;
        border-bottom: 10px solid $light;
    }
}

.selected, .item {
    display: block;
    padding: $offset-sm;
    border-radius: $border-width-sm;
    background: #fff;
    user-select: none;
    cursor: pointer;
}

.label {
    font-size: $font-size-sx;
    margin-bottom: $offset-sm;
}

.item {
    border-bottom: $border-width-sm solid $light;

    &:last-of-type {
        border-bottom: 0;
    }

    &:hover, &:focus {
        background: $primary;
        color: #fff;
    }
}
