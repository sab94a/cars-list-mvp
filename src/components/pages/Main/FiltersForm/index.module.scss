@import 'components/lib/variables.scss';

.root {
    padding: $offset-lg;
    border: $border-width-sm solid $light;
    text-align: right;
}

.select {
    margin-bottom: $offset-md;

    &:last-of-type {
        margin-bottom: $offset-lg;
    }
}