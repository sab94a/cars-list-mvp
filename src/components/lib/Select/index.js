// @flow

import React from 'react';
import cx from 'classnames';
import { SelectItem, Select as SelectType } from 'components/lib/types'
import styles from './index.module.scss';

export default class Select extends React.PureComponent<SelectType> {
    static placeholder = 'None';
    static clearSelectValue = '/';

    static defaultProps = {
        defaultValue: null,
        placeholder: Select.placeholder,
        options: [],
        onChange: () => null
    };

    state = {
        active: null,
        open: false
    }

    handlers = {}

    get emptyOption(){ 
        return {
            title: this.props.placeholder,
            value: Select.clearSelectValue
        }
    }

    get items() {
        return [this.emptyOption, ...this.props.options]
    };

    componentDidMount() {
        const { defaultValue } = this.props;

        if (defaultValue) {
            this.setState({
                active: this.findItem(defaultValue).title
            })
        }
    }

    findItem = (value:string) => {
        for (let item of this.props.options) {
            if (item.value === value) {
                return item;
            };
        }
        return this.emptyOption
    };

    bindClick(value: string) {
        if(!this.handlers[value]) {
            this.handlers[value] = (event: SyntheticEvent<*>) => {
                event.preventDefault();

                if (value === Select.clearSelectValue) {
                    this.props.onChange(false)
                } else {
                    this.props.onChange(value)
                }

                this.setState({
                    active: value,
                    open: false
                })
            }
        }
        return this.handlers[value]
    }

    togglePanel = (event: SyntheticEvent<*>) => {
        event.preventDefault();

        this.setState({ open: !this.state.open })
    }


    onBlur = () => {
        this.setState({ open: false })
    }

    render() {
        const { label, className } = this.props;
        const { open, active } = this.state;
        const selectedClasses = cx(styles.selected, { [styles.open]: open})

        return (
            <div className={ cx(className, styles.root) } tabIndex="0" onBlur={ this.onBlur }>
                { !!label && (
                    <div className={ styles.label }>{ label }</div> 
                )}
                <div className={ selectedClasses } onClick={ this.togglePanel }>
                    { this.findItem(active).title }
                </div>
                { open && (
                    <ul className={ styles.list}>
                        { this.items.map(({ value, title }: SelectItem) => (
                            <li 
                                key={ value } 
                                className={ styles.item }
                                onClick={ this.bindClick(value) }
                            >{ title }</li>
                        )) }
                    </ul> 
                )}
            </div>
        )
    }
}
