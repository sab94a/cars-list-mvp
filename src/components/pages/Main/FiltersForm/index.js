//@flow 

import React from 'react';
import Select from 'components/lib/Select';
import Button from 'components/lib/Button';
import styles from './index.module.scss';

export type Props = {
    color: string,
    manufacturer: string,
    colors: Array<string>,
    manufacturers: Array<string>
}

class FiltersForm extends React.PureComponent<Props> {
    static Titles = {
        Color: 'Color',
        Manufacturer: 'Manufacturer',
        Submit: 'Filter'
    }

    static Placeholders = {
        Color: 'All car colors',
        Manufacturer: 'All manufacturers'
    }

    static Fields = {
        Color: 'color',
        Manufacturer: 'manufacturer'
    }
    
    static defaultProps = {
        onSubmit: () => null
    }
    
    state = {
        color: this.props.color,
        manufacturer: this.props.manufacturer
    }

    handlers = {}

    bindChange(field) {
        if(!this.handlers[field]) {
            this.handlers[field] = value => {
                this.setState({
                    [field]: value
                })
            }
        }
        return this.handlers[field];
    };

    onSubmit = e => {
        e.preventDefault();

        this.props.onSubmit(this.state)
    };

    componentDidUpdate(prevProps) {
        // I know about this article https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#when-to-use-derived-state
        // But I think that in this expamle it's probably only one solution
        // Because we should have it's own state and we should update state, when props changed
        // For example if props changed from Url Query params

        const { color, manufacturer } = this.props;

        let state = {};

        if (color !== prevProps.color) {
            state.color = color
        };

        if (manufacturer !== prevProps.manufacturer) {
            state.manufacturer = manufacturer
        };

        if(Object.keys(state).length) {
            this.setState(state)
        };
    }

    render() {
        const { colors, manufacturers } = this.props;
        const { color, manufacturer } = this.state;

        return (
            <form className={ styles.root } onSubmit={ this.onSubmit }>
                <Select 
                    className={ styles.select } 
                    value={ color }
                    onChange={ this.bindChange(FiltersForm.Fields.Color) }
                    label={ FiltersForm.Titles.Color }
                    placeholder={ FiltersForm.Placeholders.Color }
                    options={ colors } 
                />
                <Select 
                    className={ styles.select } 
                    value={ manufacturer }
                    onChange={ this.bindChange(FiltersForm.Fields.Manufacturer) }
                    label={ FiltersForm.Titles.Manufacturer }
                    placeholder={ FiltersForm.Placeholders.Manufacturer }
                    options={ manufacturers } 
                />
                <Button type='submit'>{ FiltersForm.Titles.Submit }</Button>
            </form>
        )
    }
}

export default FiltersForm;
