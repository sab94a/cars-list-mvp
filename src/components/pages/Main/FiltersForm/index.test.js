import React from 'react';

import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Select from 'components/lib/Select';
import Button from 'components/lib/Button';

import FiltersForm from './';


describe('<FiltersForm />', () => {
    const color = 'white';
    const colors = [{
        value: 'black',
        title: 'black'
    }, {
        value: 'white',
        title: 'white'
    }];

    const manufacturer = 'value_2'
    const manufacturers = [{
        value: 'value_1',
        title: 'value_1'
    }, {
        value: 'value_2',
        title: 'value_2'
    }, {
        value: 'value_3',
        title: 'value_3'
    }];

    let onSubmit;
    let component;
    let defaultProps;

    const create = params => {
        component = mount(<FiltersForm { ...params } />);
    };

    beforeEach(() => {
        onSubmit = jest.fn();
        defaultProps = {
            color,
            colors,
            manufacturer,
            manufacturers,
            onSubmit
        }
    })

    it('Should render correctly', () => {
        create(defaultProps);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('Should render Color Select with correct params', () => {
        const $select = component.find(Select).at(0);

        expect($select.prop('value')).toBe(color);
        expect($select.prop('options')).toEqual(colors);
        expect($select.prop('label')).toBe(FiltersForm.Titles.Color);
        expect($select.prop('placeholder')).toBe(FiltersForm.Placeholders.Color);
    });

    it('Should render Manufacturer Select with correct params', () => {
        const $select = component.find(Select).at(1);

        expect($select.prop('value')).toBe(manufacturer);
        expect($select.prop('options')).toEqual(manufacturers);
        expect($select.prop('label')).toBe(FiltersForm.Titles.Manufacturer);
        expect($select.prop('placeholder')).toBe(FiltersForm.Placeholders.Manufacturer);
    });


    it('Should update FiltersForm state on Color Select change', () => {
        const $select = component.find(Select).at(0);
        const cb = $select.prop('onChange');
        const value = 'new color';

        cb(value);

        expect(component.state('color')).toBe(value);
    });

    it('Should update FiltersForm state on Manufacturer Select change', () => {
        const $select = component.find(Select).at(1);
        const cb = $select.prop('onChange');
        const value = 'new manufacturer';

        cb(value);

        expect(component.state('manufacturer')).toBe(value);
    });

    it('Should call onSubmit prop with state', () => {
        create(defaultProps);
        
        const state = {
            color: 'color',
            manufacturer: 'manufacturer'
        };
        const cb = component.find('form').prop('onSubmit');

        component.setState(state);

        cb({ preventDefault: () => null });

        expect(onSubmit).toHaveBeenCalled();
        expect(onSubmit).toHaveBeenCalledWith(state);
    })
});

