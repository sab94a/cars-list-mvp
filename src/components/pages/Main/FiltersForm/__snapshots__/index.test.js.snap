// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FiltersForm /> Should render correctly 1`] = `
<FiltersForm
  color="white"
  colors={
    Array [
      Object {
        "title": "black",
        "value": "black",
      },
      Object {
        "title": "white",
        "value": "white",
      },
    ]
  }
  manufacturer="value_2"
  manufacturers={
    Array [
      Object {
        "title": "value_1",
        "value": "value_1",
      },
      Object {
        "title": "value_2",
        "value": "value_2",
      },
      Object {
        "title": "value_3",
        "value": "value_3",
      },
    ]
  }
  onSubmit={[MockFunction]}
>
  <form
    className="root"
    onSubmit={[Function]}
  >
    <Select
      className="select"
      label="Color"
      onChange={[Function]}
      options={
        Array [
          Object {
            "title": "black",
            "value": "black",
          },
          Object {
            "title": "white",
            "value": "white",
          },
        ]
      }
      placeholder="All car colors"
      value="white"
    >
      <div
        className="select root"
        onBlur={[Function]}
        tabIndex="0"
      >
        <div
          className="label"
        >
          Color
        </div>
        <div
          className="selected"
          onClick={[Function]}
        >
          white
        </div>
      </div>
    </Select>
    <Select
      className="select"
      label="Manufacturer"
      onChange={[Function]}
      options={
        Array [
          Object {
            "title": "value_1",
            "value": "value_1",
          },
          Object {
            "title": "value_2",
            "value": "value_2",
          },
          Object {
            "title": "value_3",
            "value": "value_3",
          },
        ]
      }
      placeholder="All manufacturers"
      value="value_2"
    >
      <div
        className="select root"
        onBlur={[Function]}
        tabIndex="0"
      >
        <div
          className="label"
        >
          Manufacturer
        </div>
        <div
          className="selected"
          onClick={[Function]}
        >
          value_2
        </div>
      </div>
    </Select>
    <Memo(Button)
      type="submit"
    >
      <button
        className="root"
        type="submit"
      >
        Filter
      </button>
    </Memo(Button)>
  </form>
</FiltersForm>
`;
